overlord/snapstate/snapstate.go

```diff
@@ -1980,17 +1989,156 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
 	// return all names as potentially getting updated even though some may be
 	// held.
 	names := make([]string, len(updates))
+	toUpdate := make(map[string]*refreshCandidate, len(updates))
 	for i, up := range updates {
 		names[i] = up.InstanceName()
+		toUpdate[up.InstanceName()] = hints[up.InstanceName()]
 	}
 	sort.Strings(names)
 
-	// TODO: store the list of snaps to update on the conditional-auto-refresh task
-	// (this may be a subset refresh-candidates due to conflicts).
+	// store the list of snaps to update on the conditional-auto-refresh task
+	// (this may be a subset of refresh-candidates due to conflicts).
+	ar.Set("snaps", toUpdate)
 
 	return names, tss, nil
 }
 
+// autoRefreshPhase2 creates tasks for refreshing snaps from updates.
+func autoRefreshPhase2(st *state.State, updates []*refreshCandidate) ([]*state.TaskSet, error) {
```

Comment:
how much different is the core of this from doUpdate now, I'm not trying to force sharing, but there are methods/bits that we should drop from refreshCandidate and minimalInstallInfo if we are not going to reuse some kernel or doUpdate and need to cast here anyway